src/burn/engine/plan.cpp

```diff
@@ -1266,18 +1312,72 @@ extern "C" HRESULT PlanDefaultRelatedBundleRequestState(
             *pRequestState = BOOTSTRAPPER_REQUEST_STATE_REPAIR;
         }
         break;
-    case BOOTSTRAPPER_RELATION_DETECT:
+    case BOOTSTRAPPER_RELATED_BUNDLE_PLAN_TYPE_DOWNGRADE: __fallthrough;
+    case BOOTSTRAPPER_RELATED_BUNDLE_PLAN_TYPE_NONE:
         break;
     default:
         hr = E_UNEXPECTED;
-        ExitOnFailure(hr, "Unexpected relation type encountered during plan: %d", relatedBundleRelationType);
+        ExitOnFailure(hr, "Unexpected plan relation type encountered during plan: %d", relatedBundleRelationType);
         break;
     }
 
 LExit:
     return hr;
 }
 
+extern "C" HRESULT PlanRelatedBundlesInitialize(
+    __in BURN_USER_EXPERIENCE* pUserExperience,
+    __in BURN_REGISTRATION* pRegistration,
+    __in BOOTSTRAPPER_RELATION_TYPE relationType,
+    __in BURN_PLAN* pPlan
+    )
+{
+    HRESULT hr = S_OK;
+    BOOL fUninstalling = BOOTSTRAPPER_ACTION_UNINSTALL == pPlan->action || BOOTSTRAPPER_ACTION_UNSAFE_UNINSTALL == pPlan->action;
+
+    for (DWORD i = 0; i < pRegistration->relatedBundles.cRelatedBundles; ++i)
+    {
+        BURN_RELATED_BUNDLE* pRelatedBundle = pRegistration->relatedBundles.rgRelatedBundles + i;
+
+        pRelatedBundle->defaultRequestedRestore = BOOTSTRAPPER_REQUEST_STATE_NONE;
+        pRelatedBundle->requestedRestore = BOOTSTRAPPER_REQUEST_STATE_NONE;
+        pRelatedBundle->restore = BOOTSTRAPPER_ACTION_STATE_NONE;
+        pRelatedBundle->package.defaultRequested = BOOTSTRAPPER_REQUEST_STATE_NONE;
+        pRelatedBundle->package.requested = BOOTSTRAPPER_REQUEST_STATE_NONE;
+        pRelatedBundle->defaultPlanRelationType = BOOTSTRAPPER_RELATED_BUNDLE_PLAN_TYPE_NONE;
+        pRelatedBundle->planRelationType = BOOTSTRAPPER_RELATED_BUNDLE_PLAN_TYPE_NONE;
+
+        // Determine the plan relation type even if later it is ignored due to the planned action, the command relation type, or the related bundle not being plannable.
+        // This gives more information to the BA in case it wants to override default behavior.
+        // Doing it during plan instead of Detect allows the BA to change its mind without having to go all the way through Detect again.
+        hr = PlanDefaultRelatedBundlePlanType(pRelatedBundle->detectRelationType, pRegistration->pVersion, pRelatedBundle->pVersion, &pRelatedBundle->defaultPlanRelationType);
+        ExitOnFailure(hr, "Failed to get default plan type for related bundle.");
+
+        pRelatedBundle->planRelationType = pRelatedBundle->defaultPlanRelationType;
+
+        hr = UserExperienceOnPlanRelatedBundleType(pUserExperience, pRelatedBundle->package.sczId, &pRelatedBundle->planRelationType);
+        ExitOnRootFailure(hr, "BA aborted plan related bundle type.");
+
+        if (BOOTSTRAPPER_RELATED_BUNDLE_PLAN_TYPE_DOWNGRADE == pRelatedBundle->planRelationType &&
+            pRelatedBundle->fPlannable && !fUninstalling && BOOTSTRAPPER_RELATION_UPGRADE != relationType)
```

Comment:
I wasn't sure which way to go here. If the related bundle isn't plannable (registered in ARP, but missing from the package cache so it can't be run), should it block the current bundle from downgrading it?